paychmgr/paych.go

```diff
@@ -147,62 +93,73 @@ func (pm *Manager) GetChannelInfo(addr address.Address) (*ChannelInfo, error) {
 	return pm.store.getChannelInfo(addr)
 }
 
-// checks if the given voucher is valid (is or could become spendable at some point)
+// CheckVoucherValid checks if the given voucher is valid (is or could become spendable at some point)
 func (pm *Manager) CheckVoucherValid(ctx context.Context, ch address.Address, sv *paych.SignedVoucher) error {
+	_, err := pm.checkVoucherValid(ctx, ch, sv)
+	return err
+}
+
+func (pm *Manager) checkVoucherValid(ctx context.Context, ch address.Address, sv *paych.SignedVoucher) (*paych.State, error) {
 	act, pca, err := pm.loadPaychState(ctx, ch)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	var account account.State
 	_, err = pm.sm.LoadActorState(ctx, pca.From, &account, nil)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	from := account.Address
 
 	// verify signature
 	vb, err := sv.SigningBytes()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	// TODO: technically, either party may create and sign a voucher.
 	// However, for now, we only accept them from the channel creator.
 	// More complex handling logic can be added later
 	if err := sigs.Verify(sv.Signature, from, vb); err != nil {
-		return err
+		return nil, err
 	}
 
 	sendAmount := sv.Amount
 
-	// now check the lane state
-	// TODO: should check against vouchers in our local store too
-	// there might be something conflicting
-	ls := findLane(pca.LaneStates, uint64(sv.Lane))
-	if ls == nil {
-	} else {
-		if (ls.Nonce) >= sv.Nonce {
-			return fmt.Errorf("nonce too low")
-		}
-
-		sendAmount = types.BigSub(sv.Amount, ls.Redeemed)
+	// Check the voucher against the highest known voucher nonce / value
+	ls, err := pm.laneState(pca, ch, sv.Lane)
+	if err != nil {
+		return nil, err
+	}
+	// If there has been at least once voucher redeemed, and the voucher
+	// nonce value is less than the highest known nonce
+	if ls.Redeemed.Int64() > 0 && sv.Nonce <= ls.Nonce {
+		return nil, fmt.Errorf("nonce too low")
+	}
+	// If the voucher amount is less than the highest known voucher amount
+	if sv.Amount.LessThanEqual(ls.Redeemed) {
+		return nil, fmt.Errorf("voucher amount is lower than amount for voucher with lower nonce")
 	}
```

Comment:
There are some cases where we would want to support that, though for now at least we can keep this check

Reply:
Oh interesting.. could you give an example?

Reply:
It's how you'd handle returning funds to the channel creator. I don't think we have a use for this in filecoin right now, but it may be useful for users developing on top of our payment channels.

(We can add support for this later)

Reply:
I see - should I remove the check then?